dtls/server-dtls13-event.c

```diff
@@ -0,0 +1,511 @@
+        fprintf(stderr, "wolfSSL_dtls_set_peer error.\n");
+        goto error;
+    }
+
+    drop = 0;
+    ret = wolfSSL_accept(ssl);
+    if (ret != WOLFSSL_SUCCESS) {
+        err = wolfSSL_get_error(ssl, 0);
+        if (err != WOLFSSL_ERROR_WANT_READ) {
+            fprintf(stderr, "error = %d, %s\n", err,
+                    wolfSSL_ERR_reason_error_string(err));
+            fprintf(stderr, "SSL_accept failed.\n");
+            free_resources();
+            wolfSSL_Cleanup();
+            exit(1);
+        }
+    }
+
+error:
+    /* Drop the datagram */
+    if (drop)
+        (void)recv(listenfd, &b, sizeof(b), 0);
+}
+
+/* Called when we have verified a connection */
+static int chGoodCb(WOLFSSL* ssl, void* arg)
+{
+    int fd = INVALID_SOCKET;
+    struct sockaddr_in cliaddr;         /* the client's address */
+    socklen_t          cliLen = sizeof(cliaddr);
+    conn_ctx* new_ctx = (conn_ctx*)calloc(1, sizeof(conn_ctx));
+    int timeout = wolfSSL_dtls_get_current_timeout(ssl);
+    struct timeval tv;
+
+    (void)arg;
+
+    if (new_ctx == NULL) {
+        fprintf(stderr, "Out of memory!\n");
+        goto error;
+    }
+
+    new_ctx->ssl = ssl;
+
+    if (wolfSSL_dtls_get_peer(ssl, &cliaddr, &cliLen) != WOLFSSL_SUCCESS) {
+        fprintf(stderr, "wolfSSL_dtls_get_peer failed\n");
+        goto error;
+    }
+
+    /* Promote the pending connection to an active connection */
+    if (!newPendingSSL())
+        goto error;
+
+    /* We need to change the sfd here so that the ssl object doesn't drop any
+     * new connections */
+    fd = newFD();
+    if (fd == INVALID_SOCKET)
+        goto error;
+
+    /* Limit new SFD to only this connection */
+    if (connect(fd, (const struct sockaddr*)&cliaddr, cliLen) != 0) {
+        perror("connect()");
+        goto error;
+    }
+
+    /* Have the ssl object use only the SFD without the peer address (since the
+     * SFD is now connected) */
+    if (wolfSSL_dtls_set_peer(ssl, NULL, 0) != WOLFSSL_SUCCESS) {
+        fprintf(stderr, "wolfSSL_dtls_set_peer error.\n");
+        goto error;
+    }
+
+    if (wolfSSL_set_fd(ssl, fd) != WOLFSSL_SUCCESS) {
+        fprintf(stderr, "wolfSSL_set_fd error.\n");
+        goto error;
+    }
+
+    new_ctx->writeEv = event_new(base, fd, EV_WRITE, dataReady, new_ctx);
+    if (new_ctx->writeEv == NULL) {
+        fprintf(stderr, "event_new failed for srvEvent\n");
+        goto error;
+    }
+    new_ctx->readEv = event_new(base, fd, EV_READ, dataReady, new_ctx);
+    if (new_ctx->readEv == NULL) {
+        fprintf(stderr, "event_new failed for srvEvent\n");
+        goto error;
+    }
+    memset(&tv, 0, sizeof(tv));
+    if (wolfSSL_dtls13_use_quick_timeout(ssl)) {
+        if (timeout >= QUICK_MULT)
+            tv.tv_sec = timeout / QUICK_MULT;
+        else
+            tv.tv_usec = timeout * 1000000 / QUICK_MULT;
+    }
+    else
+        tv.tv_sec = timeout;
+    /* We are using non-blocking sockets so we will definitely be waiting for
```

Comment:
should we now call accept again? we exit out with a "fake" WANT_READ, but what we really want is to write the rest of the handshake.

Reply:
By returning `0` the `WOLFSSL` object will continue with the handshake. There is no need to call accept again because we return `WANT_READ` when we have sent a HRR to an unverified CH1.

Reply:
Yes, I was confused. We are already in `wolfSSL_Accept()`. Got it.